### PR TITLE
Lobby tests

### PR DESCRIPTION
**Pull request by damencho**

### CODE REVIEW

File: src/test/java/org/jitsi/meet/test/PreJoinTest.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+        joinSecondParticipant(meetingUrl);
+
+        PreJoinScreen preJoinScreen = getParticipant2().getPreJoinScreen();
+
+        preJoinScreen.waitForLoading();
+
+        // The event that display name is required takes a while to be received
+        // after showing the UI.
+        TestUtils.waitMillis(2000);
```
Comment (paweldomas) [inline src/test/java/org/jitsi/meet/test/PreJoinTest.java]: is it not possible to wait for something specific? maybe for the `joinButton` below until clickable

File: src/test/java/org/jitsi/meet/test/pageobjects/web/KnockingParticipantList.java [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+        List<Participant> participants = new ArrayList<>();
+
+        try
+        {
+            TestUtils.waitForDisplayedElementByID(participant.getDriver(), KNOCKING_PARTICIPANT_LIST_ID, 3);
+        }
+        catch(TimeoutException ex)
+        {
+            // if the list is missing return empty list of participants
+            Logger.getGlobal().log(Level.WARNING, "No knocking participant list ");
```
Comment (paweldomas) [inline src/test/java/org/jitsi/meet/test/pageobjects/web/KnockingParticipantList.java:82]: wasn't there a util method for logging somewhere?

File: src/test/java/org/jitsi/meet/test/pageobjects/web/KnockingParticipantList.java [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+
+        /**
+         * Clicks the action button for the current participant.
+         * @param buttonId the button Id to click.
+         */
+        private void clickButton(String buttonId)
+        {
+            List<WebElement> listElements = participant.getDriver().findElements(By.xpath(PARTICIPANTS_XPATH));
+            WebElement participantElement = listElements.stream().filter(participant -> {
+                String name = participant.findElement(ByTestId.testId(PARTICIPANT_NAME_TEST_ID)).getText();
```
Comment (paweldomas) [inline src/test/java/org/jitsi/meet/test/pageobjects/web/KnockingParticipantList.java:161]: probably it's doable with a selector, but not a big deal since it's contained in this method

File: src/test/java/org/jitsi/meet/test/pageobjects/web/SecurityDialog.java [hunk 4]
```diff
@@ -246,4 +248,68 @@ private boolean getLockStateByClass(String className)
...
+    /**
+     * Returns the switch that can be used to detect lobby state or change lobby state.
+     * @return the lobby switch UI element.
+     */
+    private WebElement getLobbySwitch()
+    {
+        WebDriver driver = participant.getDriver();
+        WebElement lobbySection = driver.findElement(By.id(LOBBY_SECTION_ID));
+
+        if (lobbySection != null)
```
Comment (paweldomas) [inline src/test/java/org/jitsi/meet/test/pageobjects/web/SecurityDialog.java:261]: findElement doesn't throw?

File: src/test/java/org/jitsi/meet/test/pageobjects/web/SecurityDialog.java [hunk 5]
```diff
@@ -246,4 +248,68 @@ private boolean getLockStateByClass(String className)
...
+     * Toggles Lobby.
+     */
+    public void toggleLobby()
+    {
+        WebDriver driver = participant.getDriver();
+        WebElement lobbySwitch = getLobbySwitch();
+
+        if (lobbySwitch == null)
+        {
+            throw new NoSuchElementException("Lobby switch not found!");
```
Comment (paweldomas) [inline src/test/java/org/jitsi/meet/test/pageobjects/web/SecurityDialog.java:293]: I thought `findElement` calls inside `getLobbySwitch()` would throw already

File: src/test/java/org/jitsi/meet/test/util/ByTestId.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+     */
+    public static By testId(String testId)
+    {
+        return new ByTestId(testId);
+    }
+
+    @Override
+    public List<WebElement> findElements(SearchContext context)
+    {
+        return context.findElements(By.xpath(".//*[@data-testid = '" + testId + "']"));
```
Comment (paweldomas) [inline src/test/java/org/jitsi/meet/test/util/ByTestId.java:56]: I think you can use something similar for finding participants by name, right now there are loops going over each and checking text

File: src/test/java/org/jitsi/meet/test/web/WebParticipant.java [hunk 7]
```diff
@@ -56,6 +56,9 @@
             + "&config.gatherStats=true"
             + "&config.disable1On1Mode=true"
             + "&config.analytics.disabled=true"
+            + "&config.disableAudioLevels=true"
```
Comment (paweldomas) [inline src/test/java/org/jitsi/meet/test/web/WebParticipant.java:59]: why this change?

File: src/test/java/org/jitsi/meet/test/PreJoinTest.java [hunk 8]
```diff
@@ -77,15 +78,21 @@ public void testPreJoinForRoomWithLobby()
...
-
         WebElement joinButton = preJoinScreen.getJoinButton();
 
         assertTrue(joinButton.isDisplayed(), "Join button not displayed");
 
-        String classes = joinButton.getAttribute("class");
-        assertTrue(classes.contains("disabled"), "The join button should be disabled");
+        // The event that display name is required takes a while to be received
+        // after showing the UI.
+        TestUtils.waitForCondition(
```
Comment (paweldomas) [inline src/test/java/org/jitsi/meet/test/PreJoinTest.java:87]: disregard my other comment about the wait, I was looking commit by commit

File: src/test/java/org/jitsi/meet/test/pageobjects/web/SecurityDialog.java [hunk 9]
```diff
@@ -188,6 +192,14 @@ public void open()
...
 
         this.clickToolbarButton();
+
+        // waits till its open
+        new WebDriverWait(participant.getDriver(), 3).until(
+            (ExpectedCondition<Boolean>) d -> isOpen());
+
+        // give time the dialog to settle before operating on it, we see failures where click button handlers are
+        // not triggered when clicked shortly after opening dialog, not always reproducible
+        TestUtils.waitMillis(500);
```
[the remark this answers is not in the document]
Comment (damencho) [reply src/test/java/org/jitsi/meet/test/pageobjects/web/SecurityDialog.java:204]: Yes, only the clickable is not enough 🤷‍♂️

File: src/test/java/org/jitsi/meet/test/pageobjects/web/KnockingParticipantList.java (hunk 2 above)
Comment (damencho) [reply src/test/java/org/jitsi/meet/test/pageobjects/web/KnockingParticipantList.java:82]: Yeah, I was thinking the same and did not find anything ...

File: src/test/java/org/jitsi/meet/test/pageobjects/web/SecurityDialog.java (hunk 4 above)
Comment (damencho) [reply src/test/java/org/jitsi/meet/test/pageobjects/web/SecurityDialog.java:261]: it throws if not found yes.

File: src/test/java/org/jitsi/meet/test/pageobjects/web/SecurityDialog.java (hunk 5 above)
Comment (damencho) [reply src/test/java/org/jitsi/meet/test/pageobjects/web/SecurityDialog.java:293]: Yep ... so probably I can remove this check ...

File: src/test/java/org/jitsi/meet/test/web/WebParticipant.java (hunk 7 above)
Comment (damencho) [reply src/test/java/org/jitsi/meet/test/web/WebParticipant.java:59]: Oh, that was a miss ... will revert it, sorry.

File: src/test/java/org/jitsi/meet/test/util/ByTestId.java (hunk 6 above)
Comment (damencho) [reply src/test/java/org/jitsi/meet/test/util/ByTestId.java:56]: Well actually its a little bit more complicated as we are checking the name and getting the button which is one level up:
```
<div
    <ul>
        <li>
            <div>
                <span data-testid='name'
            </div>
            <button .... 
```
Not sure how you can do this with xpath.

File: src/test/java/org/jitsi/meet/test/pageobjects/web/SecurityDialog.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (damencho) [reply src/test/java/org/jitsi/meet/test/pageobjects/web/SecurityDialog.java:204]: We just got failure in the lock room test, failure to remove the password after opening the security dialog, which this wait will fix :)
